pkg/metrics/metrics.go

```diff
@@ -16,45 +16,111 @@ package metrics
 
 import (
 	"context"
-	"net/http"
-	"strconv"
+	"fmt"
 
 	"github.com/operator-framework/operator-sdk/pkg/k8sutil"
 
-	"github.com/prometheus/client_golang/prometheus/promhttp"
 	v1 "k8s.io/api/core/v1"
-	"k8s.io/apimachinery/pkg/api/errors"
-	"sigs.k8s.io/controller-runtime/pkg/client"
-	"sigs.k8s.io/controller-runtime/pkg/client/config"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/intstr"
+	"k8s.io/client-go/rest"
+	crclient "sigs.k8s.io/controller-runtime/pkg/client"
 	logf "sigs.k8s.io/controller-runtime/pkg/runtime/log"
 )
 
 var log = logf.Log.WithName("metrics")
 
-// ExposeMetricsPort generate a Kubernetes Service to expose metrics port
-func ExposeMetricsPort() *v1.Service {
-	http.Handle("/"+k8sutil.PrometheusMetricsPortName, promhttp.Handler())
-	go http.ListenAndServe(":"+strconv.Itoa(k8sutil.PrometheusMetricsPort), nil)
+// PrometheusPortName defines the port name used in kubernetes deployment and service resources
+const PrometheusPortName = "metrics"
 
-	service, err := k8sutil.InitOperatorService()
+// ExposeMetricsPort creates a Kubernetes Service to expose the passed metrics port.
+func ExposeMetricsPort(ctx context.Context, port int32) (*v1.Service, error) {
```

Comment:
:+1: On passing in the context.

Will we want to change this function signature at all if https://github.com/kubernetes-sigs/controller-runtime/pull/273 gets merged? Would we go back to passing in the manager (or maybe the client directly)?

If so, I'm wondering if it would be worth anticipating that now to avoid an API change, or if we should just wait since we don't know exactly how it'll look.

Thoughts?

Reply:
> If so, I'm wondering if it would be worth anticipating that now to avoid an API change, or if we should just wait since we don't know exactly how it'll look.

Yes was thinking about that as well, but as currently we have no idea if that will get merged and how it will look like in the end, so not sure we can fully predict it and think about not breaking the API. And we will have to change Leader functions signature in the case we use from above PR, not sure it makes a difference here. So yes, most likely if that gets merged we will break the API, or just decide to leave it as is, we always have that choice.

Reply:
Good point about needing to change the leader election API as well.

In that case, I agree with waiting and breaking the API for both if necessary.